### DISCUSSION TITLE
Using susub graph bgraph to generate a complete graph by filtering a node Attribute

### DISCUSSION THREAD

**Discussion by joaoCarvalho1000** [graphology/graphology#422] [Q&A] [answered]
Hi all!

First of all, thank you for this amazing library.

I am failing to use the subgraph to get all nodes filtered by a specific attribute and then get a complete sub graph of my original graph but only with relevant nodes and edges connected with the output nodes of my filter. Is there a way to leverage the subgraph for this? do you recommend any other approach?

**Comment by Yomguithereal** [comment, marked as answer]
I am not sure I understand completely what you mean here (especially the complete sub graph part) but isn't using the sugbraph operator with a file function enough for your case?

```javascript
import {subgraph} from 'graphology-operators';

const sub = subgraph(graph, (node, attr) => {
  return attr.type === 'whatever';
});
```

Relevant docs here: https://graphology.github.io/standard-library/operators#subgraph

**Reply by joaoCarvalho1000** [reply]
Thank you for your quick answer. Subgraph will only give the nodes there that have the attribute "whatever", but I would like to not only get those nodes but a complete graph with all edges and nodes that connect to the "whatever" nodes. does that make more sense?

**Reply by Yomguithereal** [reply]
What you are describing *is* a subgraph. The subgraph is not only a subset of nodes but the induced graph containing only those nodes and the incident edges.

Beware of the term "complete graph" as it has a different meaning in graph theory. It means a graph where all possible edges do exist.

**Reply by joaoCarvalho1000** [reply]
ok, thank you for your explanation! I am going to try with the operator you pointed out!

**Reply by joaoCarvalho1000** [reply]
@Yomguithereal sorry to bother you again!

I have this very simple graph data (please see attachment) and I would like to use the subgraph method to filter the graph by node name "Los Angeles". Accordingly to the graph data, the expected result would be a subgraph with 1 node (the node id = 1) and 2 edges (the edge linking id 1 to id 2 and the edge linking id 1 to id 3).
However, the result that I am getting is a subgraph with 1 node (this one is correct) but with no edges.
Do you have any idea with I get this result and not the one I was expecting or the subgraph method does not work the way I was thanking.
Thank you for the support.

`{
    nodes: [
    {
      id: 1,
      name: 'Los Angeles',
      iata: 'LAX',
      lat: 33.942536,
      lng: -118.408075,
      zoom: 0
    },
    {
      id: 2,
      name: 'Rio de Janeiro',
      iata: 'GIG',
      lat: -22.808903,
      lng: -43.243647,
      zoom: 0
    },
    {
      id: 3,
      name: 'Lima',
      iata: 'LIM',
      lat: -12.021889,
      lng: -77.114319,
      zoom: 0
    }
],
    edges: [
    {
        from: 1,
        to: 2,
        zoom: 0
    },
    {
        from: 2,
        to: 3,
        zoom: 0
    },
    {
        from: 1,
        to: 3,
        zoom: 0
    }
    ]
}`

**Reply by Yomguithereal** [reply]
What you want is not a subgraph per se, and I need additional information from you to help you further: Do you only need nodes and all the edges incident to them (which means that you will include the target of said edges in the result, which also means the resulting graph is a [star graph](https://en.wikipedia.org/wiki/Star_(graph_theory)) if you only select one node like in your example? Or do you also need edges between the nodes incident to "Los Angeles" in your example, which makes this a specific kind of ego-network?

**Reply by joaoCarvalho1000** [reply]
I want to get Los Angeles all it's edges, the nodes those edges are connected  to and so on and so on. Does this help?

**Reply by Yomguithereal** [reply]
> Does this help?

Not really. Especially the "and so on and so on" part. There are multiple ways to interpret what you might want here. If what you want is the stargraph of Los Angeles (its nodes, its neighbors and the edges linking Los Angeles to its neighbors only), here is one way to do it:

```js
const sub = new Graph();

const LA = graph.findNode((key, attr) => attr.name === 'Los Angeles'));

sub.addNode(LA, graph.getNodeAttributes(LA));

graph.forEachNeighbor(LA, (neighbor, attr) => {
  sub.addNode(neighbor, attr);
  sub.addEdge(LA, neighbor, graph.getEdgeAttributes(LA, neighbor));
});
```

**Reply by Yomguithereal** [reply]
Now if you really want the actual subgraph of 'Los Angeles' and all its connected nodes (which, once again, is called an ego-network keeping ego), you can do it thusly:

```js
import {subgraph} from 'graphology-operators';

const LA = graph.findNode((key, attr) => attr.name === 'Los Angeles'));

const sub = subgraph(graph, (node, attr) => {
  return node === LA || graph.areNeighbors(node, LA);
});
```
